src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1106 @@
+       
+        //  enable wait state 1 optimization
+        self.enable_ws1_read_opt(true);
+        // change speed mode
+        self.set_flash_waitstate_and_readmode(48_000_000, 0, false);
+
+        //  By default the picocache ( a cache only for the flash) is turned off.
+        //  However the bootloader turns it on. I will explicitly turn it on here.
+        //  So if the bootloader changes, nothing breaks.
+        enable_picocache(true);
+        
+        self.current_state.set(FlashState::Ready);
+    }
+
+    pub fn get_page_size(&self) -> u32 {
+        FLASH_PAGE_SIZE
+    }
+
+    pub fn get_number_pages(&self) -> u32 {
+        //check clock and enable just incase
+        unsafe { pm::enable_clock(self.pb_clock); }
+        self.get_page_count()
+    }
+    
+    // Address is some raw address in flash that you want to read.
+    pub fn read(&self, address: usize, size: usize, buffer: &mut [u8]) -> i32 {
+        //enable clock incase it's off
+        unsafe { pm::enable_clock(self.ahb_clock); }
+        
+        //  check that address makes sense and buffer has room
+        if address > (self.get_flash_size() as usize) || 
+            address + size > (self.get_flash_size() as usize) || 
+            address + size < size || buffer.len() < size  {
+            // invalid flash address
+            return -1
+        }
+        
+        let mut byte : *const u8 = address as *const u8;
+        unsafe {
+            for i in 0..size {
+                buffer[i] = *byte;
+                byte = byte.offset(1); 
+            }
+        }
+        0
+    }
+
+    pub fn write_page(&self, page_num: i32, data: & [u8]) -> i32{
+        // enable clock incase it's off
+        unsafe { pm::enable_clock(self.ahb_clock); }
+        
+        // if we're not ready don't take the command.
+        if self.current_state.get() != FlashState::Ready {
+            return -1
+        }
+        
+        //  check data length is of size 'page_size'
+        if data.len() != self.get_page_size() as usize {
+            return -1
+        }
+
+        self.page_buffer.map(|value| {
+            value.clone_from_slice(&data);
+        });
+
+        self.page.set(page_num);
+        self.current_state.set(FlashState::Unlocking);
+        self.current_command.set(Command::Write);
+        self.lock_page_region(page_num, false);
+        0 
+    }
+    
+    pub fn erase_page(&self, page_num: i32) -> i32 {
+        // Enable AHB clock (incase it was off).
+        unsafe { pm::enable_clock(self.ahb_clock); }
+        if self.current_state.get() != FlashState::Ready {
+            return -1
+        }
+        
+        self.page.set(page_num);
+        self.current_state.set(FlashState::Unlocking);
+        self.current_command.set(Command::Erase);
+        self.lock_page_region(page_num, false);
+        0 
+    }
+}
+
+//  Assumes the only Hardware Interrupt enabled for the FLASHCALW is the
+//  FRDY (Flash Ready) interrupt.
```

Comment:
public comments in Rust (e.g. shown in documentation) have three forward slashes (`///`) in front of them.